File: pkg/store/question_store.go
package store

import (
	"context"
	"fmt"

	"gorm.io/gorm"

	"github.com/PineappleBond/xyncra-server/pkg/store/model"
)

// QuestionStore provides client-side Question persistence (D-125).
type QuestionStore struct {
	db *gorm.DB
}

// Upsert creates or updates a question (idempotent by ID).
func (qs *QuestionStore) Upsert(ctx context.Context, q *model.Question) error {
	if err := qs.db.WithContext(ctx).Save(q).Error; err != nil {
		return classifyError(fmt.Errorf("store: upsert question: %w", err))
	}
	return nil
}

// GetByConversation returns all questions for a conversation, ordered by creation time.
func (qs *QuestionStore) GetByConversation(ctx context.Context, convID string) ([]*model.Question, error) {
	var questions []*model.Question
	if err := qs.db.WithContext(ctx).
		Where("conversation_id = ?", convID).
		Order("created_at ASC").
		Find(&questions).Error; err != nil {
		return nil, classifyError(fmt.Errorf("store: get questions by conversation: %w", err))
	}
	return questions, nil
}

// DeleteByConversation removes all questions for a conversation.
func (qs *QuestionStore) DeleteByConversation(ctx context.Context, convID string) error {
	if err := qs.db.WithContext(ctx).
		Where("conversation_id = ?", convID).
		Delete(&model.Question{}).Error; err != nil {
		return classifyError(fmt.Errorf("store: delete questions by conversation: %w", err))
	}
	return nil
}

// DeleteByConversationTx removes all questions for a conversation within the
// given transaction.
func (qs *QuestionStore) DeleteByConversationTx(ctx context.Context, tx *gorm.DB, convID string) error {
	if err := tx.WithContext(ctx).
		Where("conversation_id = ?", convID).
		Delete(&model.Question{}).Error; err != nil {
		return classifyError(fmt.Errorf("store: delete questions by conversation: %w", err))
	}
	return nil
}

File: demo/vue-pure-admin/packages/xyncra-client-core/src/db/question-store.ts
/**
 * QuestionStore — client-side Question persistence (D-125).
 *
 * Mirrors Go QuestionStore (pkg/store/question_store.go).
 */

import type Dexie from 'dexie';
import type { Transaction } from 'dexie';

import type { XyncraDatabase } from './index';
import type { Question } from './models';

/**
 * QuestionStore provides client-side Question persistence (D-125).
 */
export class QuestionStore {
  constructor(private readonly db: XyncraDatabase) {}

  /**
   * Creates or updates a question (idempotent by id).
   */
  async upsert(q: Question): Promise<void> {
    await this.db.questions.put(q);
  }

  /**
   * Returns all questions for a conversation, ordered by creation time ascending.
   */
  async getByConversation(convID: string): Promise<Question[]> {
    const questions = await this.db.questions
      .where('conversation_id')
      .equals(convID)
      .toArray();

    // Sort by created_at ascending.
    questions.sort(
      (a, b) =>
        new Date(a.created_at).getTime() - new Date(b.created_at).getTime(),
    );

    return questions;
  }

  /**
   * Removes all questions for a conversation.
   */
  async deleteByConversation(convID: string): Promise<void> {
    await this.db.questions.where('conversation_id').equals(convID).delete();
  }

  /**
   * Removes all questions for a conversation within the given transaction.
   */
  async deleteByConversationTx(tx: Transaction, convID: string): Promise<void> {
    const table = tx.table('questions') as Dexie.Table<Question, string>;
    await table.where('conversation_id').equals(convID).delete();
  }
}
